### PR TITLE
Add FRI folding for M31

### PR DESCRIPTION
**Pull request by jeremyfelder**
This PR adds FRI folding for M31 field.

### CODE REVIEW

File: icicle/src/fri/fri.cu [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        ibutterfly(f0_px, f1_px, S::inverse(domain_ys[idx / 2]));
+        E f_prime = f0_px + alpha * f1_px;
+        auto folded_eval_idx = idx / 2;
+        folded_eval[folded_eval_idx] = folded_eval[folded_eval_idx] * alpha_sq + f_prime;
+      }
+    }
+  } // namespace
+
+  template <typename S, typename E>
+  cudaError_t fold_line(E* eval, S* domain_xs, E alpha, E* folded_eval, int n, FriConfig& cfg)
```
Comment (yshekel) [inline icicle/src/fri/fri.cu:50]: is int enough for size? for ntt we had this bug for large ntts in babybear.

File: icicle/src/fri/fri.cu [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    int num_threads = 256;
+    int num_blocks = (n / 2 + num_threads - 1) / num_threads;
+    fold_line_kernel<<<num_blocks, num_threads, 0, stream>>>(d_eval, d_domain_xs, alpha, d_folded_eval, n);
+
+    // Move folded_eval back to host if requested
+    if (!cfg.are_results_on_device) {
+      CHK_IF_RETURN(cudaMemcpyAsync(folded_eval, d_folded_eval, sizeof(E) * n / 2, cudaMemcpyDeviceToHost, stream));
+    }
+
```
Comment (yshekel) [inline icicle/src/fri/fri.cu:91]: need to release the allocated gpu memory

File: icicle/src/fri/fri.cu [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    int num_blocks = (n / 2 + num_threads - 1) / num_threads;
+    fold_circle_into_line_kernel<<<num_blocks, num_threads, 0, stream>>>(
+      d_eval, d_domain_ys, alpha, alpha_sq, d_folded_eval, n);
+
+    // Move folded_evals back to host if requested
+    if (!cfg.are_results_on_device) {
+      CHK_IF_RETURN(cudaMemcpyAsync(folded_eval, d_folded_eval, sizeof(E) * n / 2, cudaMemcpyDeviceToHost, stream));
+    }
+
+    // Sync if stream is default stream
```
Comment (yshekel) [inline icicle/src/fri/fri.cu:143]: same comment about memory and the int type for n.

File: wrappers/rust/icicle-fields/icicle-m31/Cargo.toml [hunk 4]
```diff
@@ -17,3 +17,4 @@ cmake = "0.1.50"
 [features]
 default = []
 devmode = ["icicle-core/devmode"]
+fri = []
```
Comment (yshekel) [inline wrappers/rust/icicle-fields/icicle-m31/Cargo.toml:20]: why does it default to off?

**Review by yshekel — APPROVED** [review]
looks good to me. Maybe would be nice to test against stwo rather than hardcoded values.

**Comment by jeremyfelder** [comment]
> Maybe would be nice to test against stwo rather than hardcoded values.

@yshekel These hard coded values are taken from Stwo's unit tests and are deterministic values. I can change the tests to use Stwo's crate to generate all of these values instead of using the hardcoded ones (though they will be the same). is that what you mean?

**Review by ChickenLover — APPROVED** [review]

**Pull request merged**